### ISSUE TITLE
Investigate effects/context system?

### ISSUE THREAD

**Issue by PatchMixolydic** [PatchMixolydic/laria#16]
Not sure if it'd have any application for my current use case (game scripting language), but this might be cool for using Laria as a general purpose language. I've run into a few situations where effects/contexts might be useful in the Triplicata game engine (a bump allocator, a virtual filesystem abstraction, maaayyybe passing important values through to Laria bindings?), so this might be worthwhile.

## Prior art
* [Context and capabilities in Rust](https://tmandry.gitlab.io/blog/posts/2021-12-21-context-capabilities/) proposes a context system for Rust (later discussion suggested that `context` should be used instead of `capability` since capabilities are a separate concept, referring to things like [`cap_std::AmbientAuthority`](https://docs.rs/cap-std/latest/cap_std/struct.AmbientAuthority.html)).
  * Dyon has a similar concept, [current objects](https://www.piston.rs/dyon-tutorial/current-objects.html). Current objects differ from `context`s in syntax (`~` over `with`), form (any type can be a current object, but `context`s must be explicitly declared), and usage (current object bindings seem to last until the end of the scope they're declared in rather than creating a new scope (similar to Koka, discussed below)).
* Koka is a research language that provides polymorphic effects; see [the Koka book](https://koka-lang.github.io/koka/doc/book.html#why-effects) for details.

**Comment by PatchMixolydic** [comment]
Contexts and effects have somewhat different semantics. A context/current object is sort of like a hidden argument that's automatically plumbed through your code (using a straw syntax here):

```rust
// still hashing out the details on the module system;
// bear with me...
mod dat;
mod vfs {
    pub enum Filesystem { /* ... */ }
    pub struct Vfs { /* ... */ }
    pub context vfs = &mut Vfs;
}

use root::{
    dat::Dat,
    vfs::{Filesystem, Vfs, vfs},
};

vfs fn get_text() -> string {
    vfs.read_to_string("foo.txt").unwrap()
}

vfs fn hello() {
    println!("{}", get_text());
}

fn main() {
    let mut vfs = Vfs::new();
    vfs.mount_root(FileSystem::Physical("."));
    let resources = Dat::load_from_compressed(...);
    vfs.mount(Filesystem::dat(resources), "resources");

    // can't do this yet since `main` doesn't have the `vfs` context
    // hello();

    // contexts don't live in the value namespace.
    // not certain if this should create a new scope or affect
    // the scope that contains it...
    install vfs = &mut vfs;
    // past this point, we now have the `vfs` context!
    // now we can call `vfs fn`s, such as `hello`
    hello();
}
```

Meanwhile, effects are meant to describe what a function is allowed to do. Examples of popular effects include `async` (allows `await`) and `throw` (unwind to raise exception). Unfortunately, it seems like effects tend to be implemented via stack unwinding. Effects might have some of the same problems as `throw`/`try`/`catch`, as effects are a generalization of them (but they can also resume the function that called the effect handler):

```rust
effect throws {
    fn throw[E: Error](error: E) -> never;
}

throws fn foo() -> string {
    throws::throw("the monitored program dumped her girlfriend");
    // don't have to do anything else since `throw(...)` returns `never`
}

throws fn bar() {
    println!("{}", foo());
}

fn main() {
    install throws::throw(error) {
        println!("{e}");
        abort();
    }

    bar();
}
```

Interestingly, I think contexts might be sort of equivalent to single-handler effects:

```rust
// *casually introduces a higher ranked trait bound containing a type*
context throw = for[E: Error] fn(E) -> never;

throw fn foo() -> string {
    throw("OOPSY I done it");
}

throw fn bar() {
    println!("{}", foo());
}

fn main() {
    install throw = for[E: Error] fn(err: E) {
        println!("{err}");
        abort();
        // you could imagine that this handler could return
        // something that could be used by the caller
    };

    bar();
}
```

**Comment by PatchMixolydic** [comment]
I'm currently leaning towards the `context` design with `with` blocks, though I'm not sure if `context` is the right keyword.

One feature which I'm not too sure about are generic effect parameters. [Koka supports this already](https://koka-lang.github.io/koka/doc/book.html#why-effects), and there's [an MCP for a similar feature in Rust](https://github.com/rust-lang/lang-team/issues/164).

**Comment by PatchMixolydic** [comment]
[This blog post reflecting on "Context and capabilities in Rust"](https://jam1.re/blog/thoughts-on-contexts-and-capabilities-in-rust) points out that using non-`Copy` owned values for `context`s is kinda weird, and suggests treating them like closure captures. Not too sure about this one since `with` blocks look more like bindings than captures, but I'm not too sure how else to deal with this. Perhaps moved types get plumbed through like usual, but can't be used in multiple callers:

```rust
struct NonCopy;
context foo = NonCopy;

foo fn reimu() {}
foo fn tsubakura() {}

fn main() {
    with foo = NonCopy::new() {
        reimu();
        tsubakura();
    }
}
```

```
error: use of moved context: `foo`
 --> src/main.rs:10:9
   |
 2 |     context foo = NonCopy;
   |             --- move occurs because `foo` has type `NonCopy`, which does not implement the `Copy` trait
   ...
 9 |     reimu();
   |     ------- context moved into `foo fn reimu` here
10 |     tsubakura();
   |     ^^^^^^^^^^^ value used here after move
```

Or perhaps they're just not allowed in the first place:
```rust
struct NonCopy;
context foo = Copy; //~ ERROR context types must either be `&mut T` or implement `Copy`
```

**Comment by PatchMixolydic** [comment]
Declaring effects before `fn` might cause parsing issues: just how far should we look ahead to determine if something's a function?
```rust
a b c d e f g h i j k l m n o p q r s t u v w x y z fn foo() {}
```

Allowing `ident` to introduce a function might inhibit the introduction of new items after Laria is stabilized:
```rust
effect union;
union fn x() {}

// Look at our new untagged unions!
union U {}
//^ ERROR expected keyword `fn`, got `U`
```
So where should effects be declared? Koka and [this rejected RFC for Rust](https://github.com/thehydroimpulse/rfcs/blob/8e92a9f6e30d8eb6a2cbec45185d010095c30a6e/active/0000-algebraic-effects-system.md) place them alongside return value:
```rust
effect union;
effect throw = for[E: Error] fn(e) -> never;

// Koka
fn foo() -> (union, throw) () {}
// RFC 73
fn foo() effect(union, throw) {}
// This feels a bit too close to C++'s `T _() const` syntax
fn foo() (union, throw) {}

// Return type specification is optional for these two
fn foo() -> () with union, throw {}
fn foo() with union, throw -> () {}
```

They could also be placed after the `fn` keyword:
```rust
fn(union, throw) foo() {}
// this might be hard to read
fn union throw foo() {}
```

**Comment by PatchMixolydic** [comment]
An MVP of effects (with no typeless effects) were implemented in b44e34f and 74b823d. I've changed the syntax of effect declarations slightly since my last post:
```rust
effect throw: fn[E: Error](E) -> never;
```
